File: signaling-server/internal/realtime/message.go
package realtime

import "encoding/json"

type Message struct {
	Type MessageType     `json:"type"`
	Data json.RawMessage `json:"data"`
}

type MessageType string

const (
	CreateSession MessageType = "create_session"
	JoinSession   MessageType = "join_session"
	Offer         MessageType = "offer"
	Answer        MessageType = "answer"
	IceCandidate  MessageType = "ice-candidate"
)

func (t MessageType) IsValid() bool {
	switch t {
	case CreateSession, JoinSession, Offer, Answer, IceCandidate:
		return true
	default:
		return false
	}
}

File: desktop/frontend/src/types/message.ts
export type Message = {
  type: MessageType;
  data?: Data;
};

type MessageType = "create_session" | "join_session" | "sdp" | "ice-candidate";

type Data = JoinSessionData | SDPData | ICECandidateDate;

export type JoinSessionData = {
  session_id: string;
};

export type SDPData = {
  type: string;
  sdp: string;
};

export type ICECandidateDate = {
  candidate: string;
  sdpMid: string;
  sdpMLineUndex: number;
};
